utils/src/main/java/com/cloud/utils/nio/Link.java

```diff
@@ -453,115 +449,150 @@ public static SSLContext initSSLContext(boolean isClient) throws GeneralSecurity
+                        // After closeOutbound the engine will be set to WRAP state,
+                        // in order to try to send a close message to the client.
+                        handshakeStatus = sslEngine.getHandshakeStatus();
+                        break;
                     }
-                    if (count == -1) {
-                        throw new IOException("Connection closed with -1 on reading size.");
+                    peerNetData.flip();
+                    try {
+                        result = sslEngine.unwrap(peerNetData, peerAppData);
+                        peerNetData.compact();
+                        handshakeStatus = result.getHandshakeStatus();
+                    } catch (SSLException sslException) {
+                        s_logger.error("SSL error occurred while processing unwrap data: " + sslException.getMessage());
+                        sslEngine.closeOutbound();
+                        handshakeStatus = sslEngine.getHandshakeStatus();
+                        break;
                     }
-                    in_pkgBuf.flip();
-                }
-                engResult = sslEngine.unwrap(in_pkgBuf, in_appBuf);
-                ByteBuffer tmp_pkgBuf = ByteBuffer.allocate(sslSession.getPacketBufferSize() + 40);
-                int loop_count = 0;
-                while (engResult.getStatus() == SSLEngineResult.Status.BUFFER_UNDERFLOW) {
-                    // The client is too slow? Cut it and let it reconnect
-                    if (loop_count > 10) {
-                        throw new IOException("Too many times in SSL BUFFER_UNDERFLOW, disconnect guest.");
+                    switch (result.getStatus()) {
+                        case OK:
+                            break;
+                        case BUFFER_OVERFLOW:
+                            // Will occur when peerAppData's capacity is smaller than the data derived from peerNetData's unwrap.
+                            peerAppData = enlargeBuffer(peerAppData, appBufferSize);
+                            break;
+                        case BUFFER_UNDERFLOW:
+                            // Will occur either when no data was read from the peer or when the peerNetData buffer
+                            // was too small to hold all peer's data.
+                            peerNetData = handleBufferUnderflow(sslEngine, peerNetData);
+                            break;
+                        case CLOSED:
+                            if (sslEngine.isOutboundDone()) {
+                                return false;
+                            } else {
+                                sslEngine.closeOutbound();
+                                handshakeStatus = sslEngine.getHandshakeStatus();
+                                break;
+                            }
+                        default:
+                            throw new IllegalStateException("Invalid SSL status: " + result.getStatus());
```

Comment:
This case statement is quite long.  Consider extracting to private method to increase readability/maintainability.